src/testers.jl

```diff
@@ -285,3 +269,58 @@ function _is_inferrable(f, args...; kwargs...)
         return false
     end
 end
+
+"""
+    _test_cotangent(accum_cotangent, ad_cotangent, fd_cotangent; kwargs...)
+
+Check if the cotangent `ad_cotangent` from `rrule` is consistent with `accum_tangent` and
+approximately equal to the cotangent `fd_cotangent` obtained with finite differencing.
+
+If `accum_cotangent` is `NoTangent()`, i.e., the argument was marked as non-differentiable,
+`ad_cotangent` and `fd_cotangent` should be `NoTangent()` as well.
+
+# Keyword arguments
+- If `check_inferred=true` (the default) and `ad_cotangent` is a thunk, then it is checked if
+  its content can be inferred.
+- All remaining keyword arguments are passed to `isapprox`.
+"""
+function _test_cotangent(
+    accum_cotangent,
+    ad_cotangent,
+    fd_cotangent;
+    check_inferred=true,
+    kwargs...,
+)
+    ad_cotangent isa AbstractThunk && check_inferred && _test_inferred(unthunk, ad_cotangent)
+
+    # The main test of the actual derivative being correct:
+    test_approx(ad_cotangent, fd_cotangent; kwargs...)
+    _test_add!!_behaviour(accum_cotangent, ad_cotangent; kwargs...)
+end
+
+# we marked the argument as non-differentiable
+function _test_cotangent(::NoTangent, ad_cotangent, ::NoTangent; kwargs...)
+    @test ad_cotangent isa NoTangent
+end
+function _test_cotangent(::NoTangent, ::ZeroTangent, ::NoTangent; kwargs...)
+    error(
+        "The pullback in the rrule should use NoTangent()" *
+        "rather than ZeroTangent() for non-perturbable arguments."
+    )
+end
+function _test_cotangent(
+    ::NoTangent,
+    ad_cotangent::ChainRulesCore.NotImplemented,
+    ::NoTangent;
+    kwargs...,
+)
+    # this situation can occur if a cotangent is not implemented and
+    # the default `rand_tangent` is `NoTangent`: e.g. due to having no fields
+    # the `@test_broken` below should tell them that there is an easy implementation for
+    # this case of `NoTangent()` (`@test_broken false` would be less useful!)
+    # https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/217
+    @test_broken ad_cotangent isa NoTangent
+end
+function _test_cotangent(::NoTangent, ad_cotangent, fd_cotangent; kwargs...)
+    error("cotangent obtained with finite differencing has to be NoTangent()")
```

Comment:
In what circumstances might this be hit? As a user of ChainRulesTestUtils I would find this error message quite confusing, if you give me a bit more context maybe I can suggest a rephrasing:)

Reply:
Users will never see it, it's just a dispatch version of the `@assert` statement. It can't happen since we compute `fd_cotangent`s in `test_rrule` and explicitly exclude the arguments with a `NoTangent()` tangent.

Reply:
E.g., does this suggest a bug in the finite differencing (upstream) code, or a bug in my downstream code?

Reply:
You will never see it. There's just no way you can end up here: https://github.com/JuliaDiff/ChainRulesTestUtils.jl/blob/f57fc7e827658bd4a00e79cb3c75d99ba292b490/src/testers.jl#L221-L222

So I really don't think we have to worry about downstream users, it is just a comment and a bit more explicit and easier to understand then the current `@assert` statement.